caas/kubernetes/provider/application/constraints.go

```diff
@@ -266,15 +281,34 @@ func processPodAffinity(pod *core.PodSpec, affinityLabels map[string]string) err
 	return nil
 }
 
-func configureConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+func configureCharmConstraint(pod *core.PodSpec, resourceName core.ResourceName, requestValue, limitValue string) (err error) {
 	if len(pod.Containers) == 0 {
 		return nil
 	}
-	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, value, pod.Containers[0].Resources.Requests)
-	if err != nil {
-		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, value)
+	for i, container := range pod.Containers {
+		if container.Name == "charm" {
+			pod.Containers[i].Resources.Requests, err = MergeConstraint(resourceName, requestValue, pod.Containers[i].Resources.Requests)
+			if err != nil {
+				return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, requestValue)
+			}
+			pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, limitValue, pod.Containers[i].Resources.Limits)
+			if err != nil {
+				return errors.Annotatef(err, "merging limit constraint %s=%s", resourceName, limitValue)
+			}
+			break
+		}
 	}
-	for i := range pod.Containers {
+	return nil
+}
+
+func configureWorkloadConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+	if len(pod.Containers) == 0 {
+		return nil
+	}
+	for i, container := range pod.Containers {
+		if container.Name == "charm" {
+			continue
+		}
 		pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, value, pod.Containers[i].Resources.Limits)
```

Comment:
We need to set the mem request for the first workload container

Reply:
refactored with thanks ian!

Reply:
That's what we do now - we set the mem request for the first container only. It just so happens the first container is the charm container and what we are doing is removing the charm container from the set of workload containers and treating it separately.